java/ql/src/semmle/code/java/dataflow/internal/DataFlowImpl.qll

```diff
@@ -3622,6 +3629,83 @@ private predicate pathThroughCallable(PathNodeMid mid, NodeEx out, CallContext c
   )
 }
 
+private module Subpaths {
+  /**
+   * Holds if `(arg, par, ret, out)` forms a subpath-tuple and `ret` is determined by
+   * `kind`, `sc`, `apout`, and `innercc`.
+   */
+  pragma[nomagic]
+  private predicate subpaths01(
+    PathNode arg, ParameterNode par, SummaryCtx sc, CallContext innercc, ReturnKindExt kind,
+    Node out, AccessPath apout
+  ) {
+    pathThroughCallable(arg, out, _, apout) and
+    pathIntoCallable(arg, par, _, innercc, sc, _) and
+    paramFlowsThrough(kind, innercc, sc, apout, _, unbind(arg.getConfiguration()))
+  }
+
+  /**
+   * Holds if `(arg, par, ret, out)` forms a subpath-tuple and `ret` is determined by
+   * `kind`, `sc`, `apout`, and `innercc`.
+   */
+  pragma[nomagic]
+  private predicate subpaths02(
+    PathNode arg, ParameterNode par, SummaryCtx sc, CallContext innercc, ReturnKindExt kind,
+    Node out, AccessPath apout
+  ) {
+    subpaths01(arg, par, sc, innercc, kind, out, apout) and
+    out = kind.getAnOutNode(_)
+  }
+
+  /**
+   * Holds if `(arg, par, ret, out)` forms a subpath-tuple.
+   */
+  private predicate subpaths03(
+    PathNode arg, ParameterNode par, PathNodeMid ret, Node out, AccessPath apout
+  ) {
+    exists(SummaryCtx sc, CallContext innercc, ReturnKindExt kind, ReturnNodeExt retnode |
+      subpaths02(arg, par, sc, innercc, kind, out, apout) and
+      ret.getNode() = retnode and
+      kind = retnode.getKind() and
+      innercc = ret.getCallContext() and
+      sc = ret.getSummaryCtx() and
+      ret.getConfiguration() = unbind(arg.getConfiguration()) and
+      apout = ret.getAp()
+    )
+  }
+
+  /**
+   * Holds if `(arg, par, ret, out)` forms a subpath-tuple, that is, flow through
+   * a subpath between `par` and `ret` with the connecting edges `arg -> par` and
+   * `ret -> out` is summarized as the edge `arg -> out`.
+   */
+  predicate subpaths(PathNode arg, PathNode par, PathNodeMid ret, PathNodeMid out) {
+    exists(ParameterNode p, Node o, AccessPath apout |
+      arg.getASuccessor() = par and
+      arg.getASuccessor() = out and
+      subpaths03(arg, p, ret, o, apout) and
+      par.getNode() = p and
+      out.getNode() = o and
+      out.getAp() = apout
+    )
+  }
+
+  private predicate parRetPair(PathNode p, PathNode ret) { subpaths(_, p, ret, _) }
+
+  /**
+   * Holds if `mid` can reach `ret` in a summarized subpath.
+   */
+  predicate retReach(PathNode mid, PathNode ret) {
+    mid = ret and parRetPair(_, ret)
+    or
+    exists(PathNode m |
+      retReach(m, ret) and
+      mid.getASuccessor() = m and
+      not subpaths(_, m, ret, _)
```

Comment:
It would, but there's some value in being able to calculate this entirely from the `subpaths` relation, as the path construction algorithm needs to infer a scope-like equivalence relation induced by this binary relation.
That said we can simplify it to
```
  predicate retReach(PathNode n) {
    parRetPair(_, n)
    or
    exists(PathNode mid |
      retReach(mid) and
      n.getASuccessor() = mid and
      not subpaths(_, mid, _, _)
    )
  }
```
which also has the upside of having a simpler negation body (i.e. a simple anti-join), and also still only depends on `subpaths` and the `edges`.